### ISSUE TITLE
Fix Verify Commit and Committed Seals

### ISSUE THREAD

**Issue by aesmonty** [clearmatics/autonity#202]
### Background

A consensus message in our Tendermint implementation is formatted as shown below (inherited from IBFT):

```golang
type message struct {
	Code          uint64
	Msg           []byte
	Address       common.Address
	Signature     []byte
	CommittedSeal []byte
}
```
In the snippet above `code` refers to the `step` message. If the `step` is either `prevote` or `precommit`, the second field, `Msg` would be of type `Vote`.

```golang
type Vote struct {
	Round             *big.Int
	Height            *big.Int
	ProposedBlockHash common.Hash
}
```
`Address` is the sender address.

`Signature` is the Keccak Sig over `message` (with signature field empty).

`CommittedSeal` is the Keccak Sig over `Vote`.

Verifying commits and committed seals properly is very important for the correct implementation of a BFT algorithm, as the requirement to insert a new block is having `> 2/3` committed seals from valid validators (which must be verified).

### Problem

Currently, in TDM we do verify the `Signature` field and verify that the `CommitedSeal` is signed by a correct validator.

However, we do not verify:

1.  That the signer of `Signature` and `CommitedSeal` are the same (and they need to be the same or the `precommit` would be invalid).
2. That the content (and length) being signed in `CommitedSeal` correspond to the contents of the `Vote` struct of that message.

### Solution

1. Implement a `VerifyCommit` function in TDM to compare that both signers are the same.
2. Add a check to fulfil condition 2.

For reference, [Correctness Analysis of IBFT](https://arxiv.org/pdf/1901.07160.pdf), page 17.

**Comment by aneequesafdar** [comment]
closed via #203

**Comment by aneequesafdar** [comment]
Reopening this issue because of incorrect assumptions about the committed seal logic were made. It was assumed that the commited seal is 
> `CommittedSeal` is the Keccak Sig over `Vote`

In reality, Committed seal is `SignECDSA(Keccak256(CONCAT(precommit.ProposalBlockHash, msgPrecommit)), PrivateKey)`, this is a major bug because proposal block hash can remain the same across rounds and thus committed seals will also remain same across rounds. This can lead to a malicious proposer collecting committed seals from different validators across different rounds and committing a block using committed seals from different rounds violating L49 of the tendermint specification.

The proposed solution is to use `precommitMsg.Signature` as the committed seal but we will now need to add the precommit messages to the extradata field to verify the signatures in `tendermint/backend/engine.go:verifyCommittedSeal` since precommit messages contain round and height information which is not at all included in the block header.
